### PR TITLE
Support tidbmonitor prometheus shards feature

### LINKED ISSUE #3925: Support shards configuration in tidbmonitor
## Feature Request

**Is your feature request related to a problem? Please describe:**
If single prometheus can't hold current targets metrics,user can reshard targets on multiple prometheus servers.
So we can use prometheus `modulus` configuration to implement , which take of the hash of the source label values , split scrape targets based on the number of shards.

**Describe the feature you'd like:**
Support shard configuration field.

### PR DESCRIPTION
**Pull request by mikechengwei**
### What problem does this PR solve?
Fix #3925 
### What is changed and how does it work?
Use prometheus modules implement prometheus horizontal expansion .
### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Release Notes
Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Support tidbmonitor prometheus shards feature.
```

### CODE REVIEW

**Comment by mikechengwei** [comment]
Screen Shot 2021-09-16 at 7 36 26 PM

File: examples/monitor-shards/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+Wait for Pods ready:
+
+```bash
+watch kubectl -n ${namespace} get pod
+```
+
+We will see that the scrape targets are distributed on the two shards.
+
+And we must use Thanos sidecar to query globally.
```
Comment (DanielZhangQD) [inline examples/monitor-shards/README.md:37]: Remote write should also be OK, right?

File: examples/monitor-shards/tidb-cluster.yaml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# IT IS NOT SUITABLE FOR PRODUCTION USE.
+# This YAML describes a basic TiDB cluster with minimum resource requirements,
+# which should be able to run in any Kubernetes cluster with storage support.
+apiVersion: pingcap.com/v1alpha1
+kind: TidbCluster
+metadata:
+  name: basic
+spec:
+  version: v5.0.1
```
Comment (DanielZhangQD) [inline examples/monitor-shards/tidb-cluster.yaml:9]:
```suggestion
  version: v5.2.1
```

File: examples/monitor-shards/tidb-monitor.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  - name: basic
+  prometheus:
+    baseImage: prom/prometheus
+    version: v2.18.1
+  grafana:
+    baseImage: grafana/grafana
+    version: 6.7.6
+  initializer:
+    baseImage: pingcap/tidb-monitor-initializer
+    version: v5.0.1
```
Comment (DanielZhangQD) [inline examples/monitor-shards/tidb-monitor.yaml:18]:
```suggestion
    version: v5.2.1
```

File: examples/monitor-shards/tidb-monitor.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  replicas: 1
+  shards: 2
+  clusters:
+  - name: basic
+  prometheus:
+    baseImage: prom/prometheus
+    version: v2.18.1
+  grafana:
+    baseImage: grafana/grafana
+    version: 6.7.6
```
Comment (DanielZhangQD) [inline examples/monitor-shards/tidb-monitor.yaml:15]:
```suggestion
    version: 7.5.7
```

File: pkg/monitor/monitor/monitor_manager.go [hunk 5]
```diff
@@ -460,22 +473,44 @@ func (m *MonitorManager) syncIngress(monitor *v1alpha1.TidbMonitor) error {
...
+	}
+	return nil
 }
 
 func (m *MonitorManager) syncGrafanaIngress(monitor *v1alpha1.TidbMonitor) error {
-	if monitor.Spec.Grafana == nil || monitor.Spec.Grafana.Ingress == nil {
-		return m.removeIngressIfExist(monitor, grafanaName(monitor))
+	shards := monitor.GetShards()
+	for shard := int32(0); shard < shards; shard++ {
+		if monitor.Spec.Grafana == nil || monitor.Spec.Grafana.Ingress == nil {
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/monitor_manager.go:500]: Move this out of the for loop?

File: pkg/monitor/monitor/monitor_manager.go [hunk 6]
```diff
@@ -460,22 +473,44 @@ func (m *MonitorManager) syncIngress(monitor *v1alpha1.TidbMonitor) error {
...
+		}
+	}
+	return nil
 }
 
 func (m *MonitorManager) syncGrafanaIngress(monitor *v1alpha1.TidbMonitor) error {
-	if monitor.Spec.Grafana == nil || monitor.Spec.Grafana.Ingress == nil {
-		return m.removeIngressIfExist(monitor, grafanaName(monitor))
+	shards := monitor.GetShards()
+	for shard := int32(0); shard < shards; shard++ {
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/monitor_manager.go:499]: Should multiple ingresses be created for different sts?
How would the user access the ingress?

File: pkg/monitor/monitor/template.go [hunk 7]
```diff
@@ -492,3 +493,30 @@ func RenderPrometheusConfig(model *MonitorConfigModel) (string, error) {
...
+	}
+	return append(relabelConfigs, &config.RelabelConfig{
+
+		SourceLabels: model.LabelNames{
+			"__address__",
+		},
+		Action:      config.RelabelHashMod,
+		TargetLabel: "__tmp_hash",
+		Modulus:     shard,
+	},
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/template.go:511]: This is not formatted?

File: examples/monitor-shards/README.md (hunk 1 above)
Comment (mikechengwei) [reply examples/monitor-shards/README.md:37]: Yes，that's no problem.

File: examples/monitor-shards/tidb-monitor.yaml [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  replicas: 1
+  shards: 2
+  clusters:
+  - name: basic
+  prometheus:
+    baseImage: prom/prometheus
+    version: v2.18.1
+  grafana:
+    baseImage: grafana/grafana
+    version: 7.5.7
```
Comment (DanielZhangQD) [inline examples/monitor-shards/tidb-monitor.yaml:15]: Remove this?

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by csuzhangxc — APPROVED** [review]

**Comment by csuzhangxc** [comment]
`TiDBCluster deploy tidb monitor specified shards normally` failed

**Pull request merged**